crates/chia-sdk-driver/src/primitives/datalayer/datastore_info.rs

```diff
@@ -0,0 +1,286 @@
+use crate::{
+    DelegationLayer, DelegationLayerArgs, DriverError, Layer, MerkleTree, NftStateLayer,
+    OracleLayer, SingletonLayer, SpendContext, WriterLayerArgs, DELEGATION_LAYER_PUZZLE_HASH,
+    DL_METADATA_UPDATER_PUZZLE_HASH,
+};
+use chia_protocol::{Bytes, Bytes32};
+use chia_puzzles::nft::NftStateLayerArgs;
+use clvm_traits::{ClvmDecoder, ClvmEncoder, FromClvm, FromClvmError, Raw, ToClvm, ToClvmError};
+use clvm_utils::{tree_hash, CurriedProgram, ToTreeHash, TreeHash};
+use num_bigint::BigInt;
+
+pub type StandardDataStoreLayers<M = DataStoreMetadata, I = DelegationLayer> =
+    SingletonLayer<NftStateLayer<M, I>>;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, ToClvm, FromClvm)]
+#[repr(u8)]
+#[clvm(atom)]
+pub enum HintType {
+    // 0 skipped to prevent confusion with () which is also none (end of list)
+    AdminPuzzle = 1,
+    WriterPuzzle = 2,
+    OraclePuzzle = 3,
+}
+
+impl HintType {
+    pub fn value(&self) -> u8 {
+        *self as u8
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub enum DelegatedPuzzle {
+    Admin(TreeHash),      // puzzle hash
+    Writer(TreeHash),     // inner puzzle hash
+    Oracle(Bytes32, u64), // oracle fee puzzle hash, fee amount
+}
+
+impl DelegatedPuzzle {
+    pub fn from_memos(remaining_memos: &mut Vec<Bytes>) -> Result<Self, DriverError> {
+        if remaining_memos.len() < 2 {
+            return Err(DriverError::MissingMemo);
+        }
+
+        let puzzle_type: u8 = u8::try_from(
+            BigInt::from_signed_bytes_be(
+                &remaining_memos
+                    .drain(0..1)
+                    .next()
+                    .ok_or(DriverError::InvalidMemo)?,
+            )
+            .to_u32_digits()
+            .1[0],
+        )
+        .map_err(|_| DriverError::InvalidMemo)?;
+
+        // under current specs, first value will always be a puzzle hash
+        let puzzle_hash: TreeHash = TreeHash::new(
+            remaining_memos
+                .drain(0..1)
+                .next()
+                .ok_or(DriverError::MissingMemo)?
+                .to_vec()
+                .try_into()
+                .map_err(|_| DriverError::InvalidMemo)?,
+        );
+
+        match puzzle_type {
```

Comment:
I'd say adding a function like this, you could just match on that
```rs
impl HintType {
    pub fn from_value(value: u8) -> Option<Self> {
        match value {
            1 => Some(Self::AdminPuzzle),
            2 => Some(Self::WriterPuzzle),
            3 => Some(Self::OraclePuzzle),
            _ => None,
        }
    }
}
```

Reply:
much nicer, thank you!